### DISCUSSION TITLE
Moving packages outside internal

### DISCUSSION THREAD

**Discussion by ahitacat** [RedHatInsights/yggdrasil#166] [Ideas]
# Idea
During my recent jobs on `rhc` and `echo worker`, I have notice that there are some useful packages that brings utilities and could be used for worker developers. I see this idea as a manner to provide a standardize set of tools to developers.

If we want to use this packages outside `yggdrasil` repository, the code must move outside the `internal` folder.

## Proposal
There can be several packages that should be moved outside of the `internal` directory, so instead of move all of them  into the root directory. We can use another folder to save them. There is a common structure for go projects where the pkg sub-directory can be used for this purpose.

The new structure for `yggdrasil` can be then:
 yggdrasil
.
├── build-aux
├── builddir
├── cmd
├── data
├── dbus
├── dist
├── doc
├── internal
├── ipc
├── lib
├── pkg
└── worker

And inside the pkg directory, we can find the standard libraries:

tree -dL 1
.
├── http
└── sync

**Comment by subpop** [comment]
I agree that `sync` and (to a lesser extent) `http` can be useful packages to out-of-tree clients, so moving them to a publicly importable location is a fine thing to do. However, I disagree with the usage of the `pkg` subdirectory. I consider that an anti-pattern in Go programming. While I was writing my response to this discussion, I was able to organize my thoughts into a [blog post](https://sub-pop.net/post/go-pkg-antipattern/) on the subject.

Let's just move `sync` and `http` into the root of the repository.

**Comment by jirihnidek** [comment]
What about creating special repository for yggdrasil worker (and example of echo worker)? This repository could also contain `sync` and `http` packages. Importing such package would require to use  `import "github.com/redhatinsights/ygg_worker/http"`.

I think that yggdrasil repository could contain only source code for the application. Anything that is necessary for building workers should be outside this repository. Or is there any reason why to hold everything tight together?

**Reply by subpop** [reply]
What about moving `sync` into the `worker` package? `github.com/redhatinsights/yggdrasil/worker/sync`?

**Reply by subpop** [reply]
Actually, that doesn't make sense. The `sync` package is used by `yggd` itself. i still think the top-level makes the most sense. `sync` is useful by itself, but isn't part of `yggdrasil` directly, so a sub-package under the top-level is the right place for it.
